lib/ssz_ex.ex

```diff
@@ -56,16 +62,108 @@ defmodule LambdaEthereumConsensus.SszEx do
   def decode(binary, module) when is_atom(module), do: decode_container(binary, module)
 
   @spec hash_tree_root!(boolean, atom) :: Types.root()
-  def hash_tree_root!(value, :bool), do: pack(value)
+  def hash_tree_root!(value, :bool), do: pack(value, :bool)
 
   @spec hash_tree_root!(non_neg_integer, {:int, non_neg_integer}) :: Types.root()
-  def hash_tree_root!(value, {:int, size}), do: pack(value, size)
+  def hash_tree_root!(value, {:int, size}), do: pack(value, {:int, size})
+
+  @spec hash_tree_root(list(), {:list, any, non_neg_integer}) ::
+          {:ok, Types.root()} | {:error, String.t()}
+  def hash_tree_root(list, {:list, type, size}) do
+    if variable_size?(type) do
+      # TODO
+      # hash_tree_root_list_complex_type(list, {:list, type, size}, limit)
+    else
+      packed_chunks = pack(list, {:list, type, size})
+      limit = chunk_count({:list, type, size})
+      len = length(list)
+      hash_tree_root_list_basic_type(packed_chunks, limit, len)
+    end
+  end
+
+  @spec hash_tree_root_list_basic_type(binary(), non_neg_integer, non_neg_integer) ::
+          {:ok, Types.root()} | {:error, String.t()}
+  def hash_tree_root_list_basic_type(chunks, limit, len) do
+    chunks_len = chunks |> byte_size() |> div(@bytes_per_chunk)
+
+    if chunks_len > limit do
+      {:error, "chunk size exceeds limit"}
+    else
+      root = merklelize_chunks(chunks, limit) |> mix_in_length(len)
+      {:ok, root}
+    end
+  end
+
+  @spec mix_in_length(Types.root(), non_neg_integer) :: Types.root()
+  def mix_in_length(root, len) do
+    {:ok, serialized_len} = encode_int(len, @bits_per_chunk)
+    root |> hash_nodes(serialized_len)
+  end
+
+  def merklelize_chunks(chunks, leaf_count \\ nil) do
+    chunks_len = chunks |> byte_size() |> div(@bytes_per_chunk)
+
+    if chunks_len == 1 and leaf_count == nil do
+      chunks
+    else
+      node_count = 2 * leaf_count - 1
+      interior_count = node_count - leaf_count
+      leaf_start = interior_count * @bytes_per_chunk
+      padded_chunks = chunks |> convert_to_next_pow_of_two(leaf_count)
+      buffer = <<0::size(leaf_start * @bits_per_byte), padded_chunks::bitstring>>
```

Comment:
We can leave this for the next PR, but using a binary as a buffer in Elixir [kills performance](https://stackoverflow.com/questions/49017003/elixir-replace-at-position-in-binary), in the same way using a tuple does. Each time you modify a binary, the VM has to copy it whole, so this is bound to be really slow for big enough inputs.

As a simple solution, using a list or array and then joining its values once at the end should work.